src/Chapter3.hs

```diff
@@ -480,6 +511,51 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data City = City
+    { cityCastle  :: Maybe Castle
+    , cityWall    :: Maybe Wall
+    , cityShrine  :: Shrine
+    , cityHouses  :: [House]
+    } deriving (Show)
+
+data Castle = Castle
+    { castleName :: String
+    } deriving (Show)
+
+data Wall = Wall
+    { wallHeight :: Int
+    } deriving (Show)
+
+data Shrine = Church | Library deriving (Show)
+
+data House = House
+    { houseRoom :: Int
+    } deriving (Show)
+
+mkHouse :: Int -> Maybe House
+mkHouse rooms
+    | rooms > 4 = Nothing
+    | otherwise = Just (House rooms)
```

Comment:
~Interesting approach. It is usually called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `mkHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂